pydantic_ai_slim/pydantic_ai/models/openrouter.py

```diff
@@ -0,0 +1,618 @@
+
+        if error := response.error:
+            raise ModelHTTPError(status_code=error.code, model_name=response.model, body=error.message)
+
+        return response
+
+    @override
+    def _process_thinking(self, message: chat.ChatCompletionMessage) -> list[ThinkingPart] | None:
+        assert isinstance(message, OpenRouterCompletionMessage)
+
+        if reasoning_details := message.reasoning_details:
+            return [_from_reasoning_detail(detail) for detail in reasoning_details]
+        else:
+            return super()._process_thinking(message)
+
+    @override
+    def _process_provider_details(self, response: chat.ChatCompletion) -> dict[str, Any]:
+        assert isinstance(response, OpenRouterChatCompletion)
+
+        provider_details = super()._process_provider_details(response)
+
+        provider_details['downstream_provider'] = response.provider
+        provider_details['native_finish_reason'] = response.choices[0].native_finish_reason
+
+        return provider_details
+
+    @override
+    def _map_model_response(self, message: ModelResponse) -> chat.ChatCompletionMessageParam:
+        texts: list[str] = []
+        tool_calls: list[chat.ChatCompletionMessageFunctionToolCallParam] = []
+        reasoning_details: list[dict[str, Any]] = []
+        for item in message.parts:
+            if isinstance(item, TextPart):
+                texts.append(item.content)
+            elif isinstance(item, ThinkingPart):
+                if item.provider_name == self.system:
+                    reasoning_details.append(_into_reasoning_detail(item).model_dump())
+                else:  # pragma: no cover
+                    pass
+            elif isinstance(item, ToolCallPart):
+                tool_calls.append(self._map_tool_call(item))
+            elif isinstance(item, BuiltinToolCallPart | BuiltinToolReturnPart):  # pragma: no cover
+                pass
+            elif isinstance(item, FilePart):  # pragma: no cover
+                pass
+            else:
+                assert_never(item)
+        message_param = chat.ChatCompletionAssistantMessageParam(role='assistant')
+        if texts:
+            message_param['content'] = '\n\n'.join(texts)
+        else:
+            message_param['content'] = None
+        if tool_calls:
+            message_param['tool_calls'] = tool_calls
+        if reasoning_details:
+            message_param['reasoning_details'] = reasoning_details  # type: ignore[reportGeneralTypeIssues]
+        return message_param
+
+    @property
+    @override
+    def _streamed_response_cls(self):
+        return OpenRouterStreamedResponse
+
+    @override
+    def _map_finish_reason(  # type: ignore[reportIncompatibleMethodOverride]
+        self, key: Literal['stop', 'length', 'tool_calls', 'content_filter', 'error']
+    ) -> FinishReason | None:
+        return _CHAT_FINISH_REASON_MAP.get(key)
+
+
+class OpenRouterChoiceDelta(chat_completion_chunk.ChoiceDelta):
```

Comment:
Please review all the classes we define here and see if they absolutely need to be public; I don't think that's the case for this one.